File: leetcode_ly/java/Tree/HouseRobberThree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int rob(TreeNode root) {
        HashMap<TreeNode,Integer> map = new HashMap<TreeNode,Integer>();
        return robber(root,map);
    }
    
    public int robber(TreeNode root,HashMap<TreeNode,Integer> map){
        if(root == null) return 0;
        
        int val = 0;
        if(map.containsKey(root)) return map.get(root);
        if(root.left != null){
            val = val + robber(root.left.left,map) + robber(root.left.right,map);
        }
        if(root.right != null){
            val = val + robber(root.right.left,map) + robber(root.right.right,map);
        }
        
        int val1 = val + root.val;
        int val2 = robber(root.left,map) + robber(root.right,map);
        int value = Math.max(val1,val2);
        map.put(root,value);
        return value;
    }
}

File: leetcode_ly/c++/Tree/HouseRobberThree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int rob(TreeNode* root) {
        map<TreeNode*,int> myMap;
        return getResult(root,myMap);
    }
    
    int getResult(TreeNode* root,map<TreeNode*,int>& myMap){
        if(!root) return 0;
        if(myMap.find(root) != myMap.end()) return myMap[root];
        
        int val = 0;
        if(root->left){
            val = val + getResult(root->left->left,myMap) + getResult(root->left->right,myMap);
        }
        if(root->right){
            val = val + getResult(root->right->left,myMap) + getResult(root->right->right,myMap);
        }
        int val1 = getResult(root->left,myMap) + getResult(root->right,myMap);
        int val2 = val + root->val;
        int result = (val2 > val1) ? val2 : val1;
        myMap[root] = result;
        return result;
    }
};